ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/GetBlockFromPeersTask.java

```diff
@@ -0,0 +1,99 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.eth.manager.task;
+
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.core.Hash;
+import org.hyperledger.besu.ethereum.eth.manager.EthContext;
+import org.hyperledger.besu.ethereum.eth.manager.EthPeer;
+import org.hyperledger.besu.ethereum.eth.manager.exceptions.IncompleteResultsException;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/** Downloads a block from a peer. Will complete exceptionally if block cannot be downloaded. */
+public class GetBlockFromPeersTask extends AbstractEthTask<AbstractPeerTask.PeerTaskResult<Block>> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final List<EthPeer> peers;
+  private final List<EthPeer> triedAndInvalidPeers;
+  private final EthContext ethContext;
+  private final ProtocolSchedule protocolSchedule;
+  private final Hash hash;
+  private final long blockNumber;
+  private final MetricsSystem metricsSystem;
+
+  protected GetBlockFromPeersTask(
+      final List<EthPeer> peers,
+      final ProtocolSchedule protocolSchedule,
+      final EthContext ethContext,
+      final Hash hash,
+      final long blockNumber,
+      final MetricsSystem metricsSystem) {
+    super(metricsSystem);
+    this.peers = peers;
+    this.triedAndInvalidPeers = new ArrayList<>();
+    this.ethContext = ethContext;
+    this.blockNumber = blockNumber;
+    this.metricsSystem = metricsSystem;
+    this.protocolSchedule = protocolSchedule;
+    this.hash = hash;
+  }
+
+  public static GetBlockFromPeersTask create(
+      final List<EthPeer> peers,
+      final ProtocolSchedule protocolSchedule,
+      final EthContext ethContext,
+      final Hash hash,
+      final long blockNumber,
+      final MetricsSystem metricsSystem) {
+    return new GetBlockFromPeersTask(
+        peers, protocolSchedule, ethContext, hash, blockNumber, metricsSystem);
+  }
+
+  @Override
+  protected void executeTask() {
+    LOG.debug("Downloading block {} from peers {}.", hash, peers.stream().map(EthPeer::toString));
+    final Optional<EthPeer> peer =
+        peers.stream().filter(ethPeer -> !triedAndInvalidPeers.contains(ethPeer)).findFirst();
+    if (peer.isPresent() && !peer.get().isDisconnected()) {
```

Comment:
If even one peer is disconnected then we stop trying for others as well? I'm working on making this method simpler anyways, I'll show you my changes once I write them up.

Reply:
Just updated the code thanks to your PR. thanks